File: app/frontend/modules/admin/modules/client/clientSsoSettings.ts
import * as t from 'io-ts'

export const ClientSsoSettingsTR = t.type({
  id: t.string,
  ssoEnabled: t.boolean,
  ssoEnforced: t.boolean,
  enforceFor: t.union([t.literal('none'), t.literal('all'), t.literal('specific_domains')]),
  enforcedDomains: t.array(t.string),
  idpEntityId: t.union([t.string, t.null]),
  idpSsoUrl: t.union([t.string, t.null]),
  idpSloUrl: t.union([t.string, t.null]),
  idpCert: t.union([t.string, t.null]),
  sessionTimeout: t.union([t.number, t.null]),
  allowedDomains: t.array(t.string),
  assertionConsumerServiceUrl: t.string,
  issuer: t.string,
})

export type ClientSsoSettings = t.TypeOf<typeof ClientSsoSettingsTR>

File: spec/factories/client_sso_settings.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :client_sso_setting do
    association :client, factory: :tenancy
    sso_enabled { false }
    sso_enforced { false }
    idp_entity_id { nil }
    idp_sso_url { nil }
    idp_slo_url { nil }
    idp_cert { nil }
    session_timeout { nil }
    allowed_domains { [] }

    trait :enabled do
      sso_enabled { true }
      idp_entity_id { "https://idp.example.com/#{SecureRandom.hex(4)}" }
      idp_sso_url { 'https://idp.example.com/sso/saml' }
      idp_slo_url { 'https://idp.example.com/sso/slo' }
      idp_cert { Rails.root.join('spec/fixtures/files/cert.pem').read }
    end

    trait :enforced do
      enabled
      sso_enforced { true }
    end
  end
end
